File: scaling-controller/solace_autoscale_client/routing_library.py
"""Trusted application-registered business routing evaluators for managed messaging."""

from __future__ import annotations

from collections.abc import Callable, Iterator, Mapping
from dataclasses import dataclass
from typing import Any, Literal, Protocol, TypeAlias

MAX_ROUTING_KEY_BYTES = 1024
MAX_HEADERS = 64
MAX_HEADER_NAME_BYTES = 128
MAX_HEADER_VALUE_BYTES = 4096
MAX_HEADERS_BYTES = 32768


class MessageView(Protocol):
    """Whole borrowed message passed to an evaluator before JSON serialization.

    The view and its payload/headers are application-owned objects. Evaluators must treat them as
    read-only and must not retain them after returning. Python cannot enforce immutability here.
    """

    topic: str
    headers: Mapping[str, str]
    payload: Any
    event_id: str


@dataclass(frozen=True)
class BusinessKey:
    """UTF-8 business key hashed with the existing sha256-json-v1 partition contract."""

    value: str


@dataclass(frozen=True)
class SHA256Digest:
    """Exact 32-byte digest mapped directly, big-endian, modulo partition count."""

    value: bytes

    @classmethod
    def from_hex(cls, value: str) -> SHA256Digest:
        if not isinstance(value, str) or len(value) != 64:
            raise RoutingEvaluationError("SHA-256 hex result must contain exactly 64 lowercase hex digits")
        try:
            raw = bytes.fromhex(value)
        except ValueError as exc:
            raise RoutingEvaluationError(
                "SHA-256 hex result must contain exactly 64 lowercase hex digits"
            ) from exc
        if value != raw.hex():
            raise RoutingEvaluationError("SHA-256 hex result must use canonical lowercase hex")
        return cls(raw)


RoutingResult: TypeAlias = BusinessKey | SHA256Digest


class RoutingEvaluator(Protocol):
    def __call__(self, message: MessageView) -> RoutingResult: ...


@dataclass(frozen=True)
class EvaluatorIdentity:
    name: str
    version: str


class RoutingEvaluationError(ValueError):
    """A deterministic local evaluator rejected a publication or delivered message."""


@dataclass(frozen=True)
class ResolvedRouting:
    kind: Literal["key", "sha256"]
    value: str
    partition: int


class RoutingEvaluatorRegistry:
    """Explicit registry of trusted evaluator code installed with the application."""

    def __init__(self) -> None:
        self._evaluators: dict[EvaluatorIdentity, RoutingEvaluator] = {}

    def register(
        self, name: str, version: str, evaluator: RoutingEvaluator
    ) -> RoutingEvaluatorRegistry:
        identity = validate_identity(name, version)
        if not callable(evaluator):
            raise TypeError("routing evaluator must be callable")
        if identity in self._evaluators:
            raise ValueError(f"routing evaluator {name!r} version {version!r} is already registered")
        self._evaluators[identity] = evaluator
        return self

    def require(self, name: object, version: object) -> RoutingEvaluator:
        identity = validate_identity(name, version)
        try:
            return self._evaluators[identity]
        except KeyError as exc:
            raise ValueError(
                f"required routing evaluator {name!r} version {version!r} is not registered"
            ) from exc

    def identities(self) -> Iterator[EvaluatorIdentity]:
        return iter(self._evaluators)

    def evaluate(
        self,
        name: str,
        version: str,
        result_format: Literal["key", "sha256"],
        message: MessageView,
        *,
        shard: str,
        partitions: int,
        key_partition: Callable[[str], int],
    ) -> ResolvedRouting:
        evaluator = self.require(name, version)
        try:
            result = evaluator(message)
        except Exception as exc:
            raise RoutingEvaluationError(
                f"routing evaluator {name!r} version {version!r} failed"
            ) from exc
        if result_format == "key":
            if not isinstance(result, BusinessKey):
                raise RoutingEvaluationError("routing evaluator must return BusinessKey for result: key")
            value = validate_key(result.value)
            return ResolvedRouting("key", value, key_partition(value))
        if result_format == "sha256":
            if not isinstance(result, SHA256Digest):
                raise RoutingEvaluationError(
                    "routing evaluator must return SHA256Digest for result: sha256"
                )
            digest = validate_digest(result.value)
            return ResolvedRouting(
                "sha256", digest.hex(), int.from_bytes(digest, "big") % partitions
            )
        raise RoutingEvaluationError(f"unsupported routing evaluator result {result_format!r}")


def validate_identity(name: object, version: object) -> EvaluatorIdentity:
    if not isinstance(name, str) or not isinstance(version, str):
        raise ValueError("routing evaluator name/version must be strings")
    if (
        not 1 <= len(name.encode("utf-8")) <= 128
        or not 1 <= len(version.encode("utf-8")) <= 64
        or "\x00" in name
        or "\x00" in version
    ):
        raise ValueError("routing evaluator name/version must be bounded nonempty UTF-8 strings")
    return EvaluatorIdentity(name, version)


def validate_key(key: object) -> str:
    if not isinstance(key, str):
        raise RoutingEvaluationError("BusinessKey value must be a string")
    size = len(key.encode("utf-8"))
    if not 1 <= size <= MAX_ROUTING_KEY_BYTES or "\x00" in key:
        raise RoutingEvaluationError(
            f"routing key must be 1-{MAX_ROUTING_KEY_BYTES} UTF-8 bytes without NUL"
        )
    return key


def validate_digest(value: object) -> bytes:
    if not isinstance(value, bytes) or len(value) != 32:
        raise RoutingEvaluationError("SHA256Digest value must contain exactly 32 bytes")
    return value


def validate_headers(headers: Mapping[str, str]) -> None:
    if len(headers) > MAX_HEADERS:
        raise ValueError(f"message headers exceed {MAX_HEADERS} entries")
    total = 0
    for name, value in headers.items():
        if not isinstance(name, str) or not isinstance(value, str):
            raise ValueError("message headers must contain string names and values")
        name_size, value_size = len(name.encode("utf-8")), len(value.encode("utf-8"))
        if not 1 <= name_size <= MAX_HEADER_NAME_BYTES or "\x00" in name:
            raise ValueError("message header name is empty, too long, or contains NUL")
        if value_size > MAX_HEADER_VALUE_BYTES or "\x00" in value:
            raise ValueError("message header value is too long or contains NUL")
        total += name_size + value_size
    if total > MAX_HEADERS_BYTES:
        raise ValueError(f"message headers exceed {MAX_HEADERS_BYTES} UTF-8 bytes")


EvaluatorFunction = Callable[[MessageView], RoutingResult]

File: shim/messaging/routing_library.go
package messaging

import (
	"errors"
	"fmt"
	"math/big"
	"sync"
	"unicode/utf8"
)

const (
	maxRoutingKeyBytes  = 1024
	maxHeaders          = 64
	maxHeaderNameBytes  = 128
	maxHeaderValueBytes = 4096
	maxHeadersBytes     = 32768
)

// Publication is borrowed by a local routing evaluator before JSON serialization.
// The caller retains ownership of Payload and Headers, must not mutate them concurrently,
// and evaluators must not mutate or retain them after Evaluate returns.
type Publication struct {
	Topic   string
	Payload any
	EventID string
	Headers map[string]string
}

// RoutingResult is intentionally sealed so a hex-looking BusinessKey can never be
// mistaken for a digest. Use BusinessKey or SHA256Digest explicitly.
type RoutingResult interface{ routingResult() }

type BusinessKey string

func (BusinessKey) routingResult() {}

type SHA256Digest [32]byte

func (SHA256Digest) routingResult() {}

// Evaluator is trusted customer code installed and registered by the application.
type Evaluator interface {
	Evaluate(*Publication) (RoutingResult, error)
}

// EvaluatorFunc adapts a function to Evaluator.
type EvaluatorFunc func(*Publication) (RoutingResult, error)

func (f EvaluatorFunc) Evaluate(m *Publication) (RoutingResult, error) { return f(m) }

type evaluatorIdentity struct{ name, version string }

// EvaluatorRegistry contains only explicitly registered, in-process implementations.
type EvaluatorRegistry struct {
	mu         sync.RWMutex
	evaluators map[evaluatorIdentity]Evaluator
}

func NewEvaluatorRegistry() *EvaluatorRegistry {
	return &EvaluatorRegistry{evaluators: map[evaluatorIdentity]Evaluator{}}
}

func (r *EvaluatorRegistry) Register(name, version string, evaluator Evaluator) error {
	id, err := validateEvaluatorIdentity(name, version)
	if err != nil {
		return err
	}
	if evaluator == nil {
		return errors.New("routing evaluator is required")
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.evaluators == nil {
		r.evaluators = map[evaluatorIdentity]Evaluator{}
	}
	if _, exists := r.evaluators[id]; exists {
		return fmt.Errorf("routing evaluator %q version %q is already registered", name, version)
	}
	r.evaluators[id] = evaluator
	return nil
}

func (r *EvaluatorRegistry) require(name, version string) (Evaluator, error) {
	id, err := validateEvaluatorIdentity(name, version)
	if err != nil {
		return nil, err
	}
	if r == nil {
		return nil, fmt.Errorf("required routing evaluator %q version %q is not registered", name, version)
	}
	r.mu.RLock()
	defer r.mu.RUnlock()
	evaluator := r.evaluators[id]
	if evaluator == nil {
		return nil, fmt.Errorf("required routing evaluator %q version %q is not registered", name, version)
	}
	return evaluator, nil
}

func validateEvaluatorIdentity(name, version string) (evaluatorIdentity, error) {
	if !utf8.ValidString(name) || !utf8.ValidString(version) || len(name) < 1 || len(name) > 128 || len(version) < 1 || len(version) > 64 || containsNUL(name) || containsNUL(version) {
		return evaluatorIdentity{}, errors.New("routing evaluator name/version must be bounded nonempty UTF-8 strings without NUL")
	}
	return evaluatorIdentity{name, version}, nil
}

func containsNUL(value string) bool {
	for _, b := range []byte(value) {
		if b == 0 {
			return true
		}
	}
	return false
}

func validateHeaders(headers map[string]string) error {
	if len(headers) > maxHeaders {
		return fmt.Errorf("message headers exceed %d entries", maxHeaders)
	}
	total := 0
	for name, value := range headers {
		if !utf8.ValidString(name) || !utf8.ValidString(value) || len(name) < 1 || len(name) > maxHeaderNameBytes || len(value) > maxHeaderValueBytes || containsNUL(name) || containsNUL(value) {
			return errors.New("message header name/value is invalid, too long, or contains NUL")
		}
		total += len(name) + len(value)
	}
	if total > maxHeadersBytes {
		return fmt.Errorf("message headers exceed %d UTF-8 bytes", maxHeadersBytes)
	}
	return nil
}

func invokeEvaluator(e Evaluator, message *Publication) (result RoutingResult, err error) {
	defer func() {
		if recover() != nil {
			err = errors.New("routing evaluator panicked")
		}
	}()
	result, err = e.Evaluate(message)
	if err != nil {
		return nil, errors.New("routing evaluator failed")
	}
	return result, nil
}

func resolveEvaluator(e Evaluator, format string, message *Publication, shard string, partitions int) (kind, value string, partition int, err error) {
	result, err := invokeEvaluator(e, message)
	if err != nil {
		return "", "", 0, err
	}
	switch format {
	case "key":
		key, ok := result.(BusinessKey)
		if !ok {
			return "", "", 0, errors.New("routing evaluator must return BusinessKey for result key")
		}
		value = string(key)
		if !utf8.ValidString(value) || len(value) < 1 || len(value) > maxRoutingKeyBytes || containsNUL(value) {
			return "", "", 0, fmt.Errorf("routing key must be 1-%d UTF-8 bytes without NUL", maxRoutingKeyBytes)
		}
		return "key", value, partitionFor(shard, value, partitions), nil
	case "sha256":
		digest, ok := result.(SHA256Digest)
		if !ok {
			return "", "", 0, errors.New("routing evaluator must return SHA256Digest for result sha256")
		}
		value = fmt.Sprintf("%x", digest[:])
		partition = int(new(big.Int).Mod(new(big.Int).SetBytes(digest[:]), big.NewInt(int64(partitions))).Int64())
		return "sha256", value, partition, nil
	default:
		return "", "", 0, errors.New("unsupported routing evaluator result")
	}
}
